File: online_judge/leetcode/binary-tree-inorder-traversal.cc
#include <vector>
#include <stack>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        helper(root, res);
        return res;
    }

    void helper(TreeNode* root, vector<int> &res) {
        if (!root) return;
        helper(root->left, res);
        res.push_back(root->val);
        helper(root->right, res);
    }

    // iterative
    vector<int> inorderTraversal2(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;

        while(root) {
            if (root->left) {
                s.push(root);
                root = root->left;
            } else {
                res.push_back(root->val);
                while(!root->right && !s.empty()) {
                    root = s.top();
                    s.pop();
                    res.push_back(root->val);
                }
                root = root->right;
            }
        }
        return res;
    }
};

File: online_judge/lintcode/binary_tree_inorder_traversal.py
"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


class Solution:
    """
    @param root: The root of binary tree.
    @return: Inorder in ArrayList which contains node values.
    """
    def inorderTraversal(self, root):
        ret = []
        self.helper(root, ret)
        return ret

    def helper(self, root, ret):
        if root:
            self.helper(root.left, ret)
            ret.append(root.val)
            self.helper(root.right, ret)
